### DISCUSSION TITLE
Function in CNS_derive.cpp

### DISCUSSION THREAD

**Discussion by wazedxwxx** [AMReX-Codes/amrex#2351] [Q&A] [answered]
I am learning the example in GPU->CNS. But the function 

void cns_derpres (const Box& bx, FArrayBox& pfab, int dcomp, int /*ncomp*/,
                  const FArrayBox& rhoefab, const Geometry& /*geomdata*/,
                  Real /*time*/, const int* /*bcrec*/, int /*level*/)
{
    auto const rhoe = rhoefab.array();
    auto       p    = pfab.array();
    Parm const* parm = CNS::d_parm;
    amrex::ParallelFor(bx,
    [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
    {
        p(i,j,k,dcomp) = (parm->eos_gamma-1.)*rhoe(i,j,k);
    });
}

void cns_dervel (const Box& bx, FArrayBox& velfab, int dcomp, int /*ncomp*/,
                  const FArrayBox& datfab, const Geometry& /*geomdata*/,
                  Real /*time*/, const int* /*bcrec*/, int /*level*/)
{
    auto const dat = datfab.array();
    auto       vel = velfab.array();
    amrex::ParallelFor(bx,
    [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
    {
        vel(i,j,k,dcomp) = dat(i,j,k,1)/dat(i,j,k,0);
    });
}

in the file CNS_derive.cpp makes me unable to understand.
I think the operations in the function body of these two functions do not seem to be executed. Even if I delete the code inside the function body, I can still get the results complied by nvcc that look correct. I want to figure out what these two functions do？

**Comment by WeiqunZhang** [comment, marked as answer]
These functions are used to compute so-called derived variables. You don't seem to see a difference by deleting the function body, because they are only used in writing plotfiles.  They compute pressure and velocity based on conservative variables.  I assume what `p = (gamma-1)*rhoe` does is obvious.  In `cns_dervel`, `dat(i,j,k,1)` is `rho*vx` and `dat(i,j,k,0)` is `rho` as specified in `CNS_setup.cpp`
```
    // Velocities                                                                                                                                                                                          
    derive_lst.add("velocity",IndexType::TheCellType(),AMREX_SPACEDIM,
                   {AMREX_D_DECL("x_velocity","y_velocity","z_velocity")},
                   cns_dervel,the_same_box);
    derive_lst.addComponent("velocity",desc_lst,State_Type,Density,1+AMREX_SPACEDIM);
```
So `vel(i,j,k,dcomp) = dat(i,j,k,1)/dat(i,j,k,0);` is just `vx = rho*vx / rho`.
